src/protocol/boolean/solved_bits.rs

```diff
@@ -0,0 +1,285 @@
+use super::bitwise_lt::BitwiseLessThan;
+use crate::error::Error;
+use crate::ff::{Field, Int};
+use crate::helpers::Role;
+use crate::protocol::boolean::BitOpStep;
+use crate::protocol::modulus_conversion::convert_shares::{ConvertShares, XorShares};
+use crate::protocol::reveal::Reveal;
+use crate::protocol::{context::ProtocolContext, RecordId};
+use crate::secret_sharing::Replicated;
+use futures::future::try_join_all;
+use std::iter::repeat;
+
+#[allow(dead_code)]
+#[derive(Debug)]
+pub struct RandomBitsShare<F: Field> {
+    b_b: Vec<Replicated<F>>,
+    b_p: Replicated<F>,
+}
+
+/// This protocol tries to generate a sequence of uniformly random sharing of
+/// bits in `F_p`. Adding these 3-way secret-sharing will yield the secret
+/// `b_i ∈ {0,1}`. This protocol will abort and returns `None` if the secret
+/// number from randomly generated bits is not less than the field's prime
+/// number. Once aborted, the caller must provide a new narrowed context if
+/// they wish to call this protocol again for the same `record_id`.
+///
+/// This is an implementation of "3.1 Generating random solved BITS" from I. Damgård
+/// et al., but replaces `RAN_2` with our own PRSS implementation in lieu.
+///
+/// 3.1 Generating random solved BITS
+/// "Unconditionally Secure Constant-Rounds Multi-party Computation for Equality, Comparison, Bits, and Exponentiation"
+/// I. Damgård et al.
+pub struct SolvedBits {}
+
+impl SolvedBits {
+    // Try generating random sharing of bits, `[b]_B`, and `l`-bit long.
+    // Each bit has a 50% chance of being a 0 or 1, so there are
+    // `F::Integer::MAX - p` cases where `b` may become larger than `p`.
+    // With `Fp32BitPrime` (prime is `2^32 - 5`), that chance is around
+    // 1 * 10^-9. For Fp31, the chance of this protocol succeeding is around
+    // 1/2^3 =~ 13% (3 high bits being all 0's).
```

Comment:
Comment out of date. New chance of failure is 1 out of 32 (~3.125%)

Reply:
I'll include this in the next PR.